irksome/imex.py

```diff
@@ -304,3 +307,225 @@ def solver_stats(self):
                 self.num_linear_iterations_prop,
                 self.num_nonlinear_iterations_it,
                 self.num_linear_iterations_it)
+
+
+def getFormsIMEX(F, Fexp, ks, khats, butch, t, dt, u0, bcs=None):
+    if bcs is None:
+        bcs = []
+
+    v = F.arguments()[0]
+    V = v.function_space()
+    msh = V.mesh()
+    assert V == u0.function_space()
+
+    num_fields = len(V)
+    num_stages = butch.num_stages
+    k = Function(V)
+    g = Function(V)
+
+    khat = Function(V)
+    ghat = Function(V)
+    vhat = TestFunction(V)
+
+    # If we're on a mixed problem, we need to replace pieces of the
+    # solution.  Stores array of the splittings of the functions for each stage.
+    if num_fields == 1:
+        k_bits = [k]
+        u0bits = [u0]
+        gbits = [g]
+        ghat_bits = [ghat]
+    else:
+        k_bits = np.array(split(k), dtype=object)
+        u0bits = split(u0)
+        gbits = split(g)
+        ghat_bits = split(g)
+
+    # Note: the Constant c is used for substitution in both the
+    # implicit variational form and BC's, and we update it for each stage in
+    # the loop over stages in the advance method.  The Constants a and chat are
+    # used similarly in the variational forms
+    MC = MeshConstant(msh)
+    c = MC.Constant(1.0)
+    chat = MC.Constant(1.0)
+    a = MC.Constant(1.0)
+
+    # Implicit replacement, solve at time t + c * dt, for k
+    repl = {t: t + c * dt}
+    for u0bit, kbit, gbit in zip(u0bits, k_bits, gbits):
+        repl[u0bit] = gbit + dt * a * kbit
+        repl[TimeDerivative(u0bit)] = kbit
+    stage_F = replace(F, repl)
+
+    # Explicit replacement, solve at time t + chat * dt, for khat
+    replhat = {t: t + chat * dt}
+    for u0bit, ghatbit in zip(u0bits, ghat_bits):
+        replhat[u0bit] = ghatbit
+    Fhat = inner(khat, vhat)*dx - replace(Fexp, replhat)
+
+    bcnew = []
+
+    # For the DIRK-IMEX case, we need one new BC for each old one
+    # (rather than one per stage), but we need a `Function` inside of
+    # each BC and a rule for computing that function at each time for
+    # each stage.
+
+    a_vals = np.array([MC.Constant(0) for i in range(num_stages)],
```

Comment:
Pablo had done something to get these as a constant array rather than array of constants, which is supposed to improve code gen time.  Check out the galerkin steppers on your branch (which I need to finally merge!)  It doesn't look like this trick is propagated through the RK stuff, so maybe we want to let this ride and come back to RK stuff in a separate PR.

Reply:
I saw that change on the Galerkin branch, but wasn't sure.  I agree - let's fix that everywhere in a separate PR.

Reply:
What is the purpose of an array of `num_stages`  zeros (wrapped as constant), anyway.  Do these get updated somewhere?

Reply:
And the key idea is putting a numpy array inside of `Constant` rather than vectorizing `Constant`.  It's pretty simple.

Reply:
N/M, I see the updates.  They get re-set in the loop inside of `advance`.  Also, this is just a vector, not a matrix, so may not be a major cause for alarm.